sdk/iot/provisioning/src/az_iot_provisioning_client.c

```diff
@@ -47,161 +54,420 @@ AZ_NODISCARD az_result az_iot_provisioning_client_init(
 }
 
 // <id_scope>/registrations/<registration_id>/api-version=<service_version>
-AZ_NODISCARD az_result az_iot_provisioning_client_user_name_get(
+AZ_NODISCARD az_result az_iot_provisioning_client_get_user_name(
     az_iot_provisioning_client const* client,
-    az_span mqtt_user_name,
-    az_span* out_mqtt_user_name)
+    char* mqtt_user_name,
+    size_t mqtt_user_name_size,
+    size_t* out_mqtt_user_name_length)
 {
   AZ_PRECONDITION_NOT_NULL(client);
-  AZ_PRECONDITION_VALID_SPAN(mqtt_user_name, 0, false);
-  AZ_PRECONDITION_NOT_NULL(out_mqtt_user_name);
+  AZ_PRECONDITION_NOT_NULL(mqtt_user_name);
+  AZ_PRECONDITION(mqtt_user_name_size > 0);
+
+  az_span mqtt_user_name_span
+      = az_span_init((uint8_t*)mqtt_user_name, (int32_t)mqtt_user_name_size);
 
   const az_span* const user_agent = &(client->_internal.options.user_agent);
+  az_span str_registrations
+      = az_span_slice(str_dps_registrations_res, idx_registrations_start, idx_registrations_end);
 
   int32_t required_length = az_span_size(client->_internal.id_scope)
       + az_span_size(str_registrations) + az_span_size(client->_internal.registration_id)
       + az_span_size(provisioning_service_api_version);
   if (az_span_size(*user_agent) > 0)
   {
-    required_length += az_span_size(*user_agent) + 1;
+    required_length += az_span_size(user_agent_version_prefix) + az_span_size(*user_agent);
   }
 
-  AZ_RETURN_IF_NOT_ENOUGH_SIZE(mqtt_user_name, required_length);
+  AZ_RETURN_IF_NOT_ENOUGH_SIZE(mqtt_user_name_span, required_length);
 
-  az_span remainder = az_span_copy(mqtt_user_name, client->_internal.id_scope);
+  az_span remainder = az_span_copy(mqtt_user_name_span, client->_internal.id_scope);
   remainder = az_span_copy(remainder, str_registrations);
   remainder = az_span_copy(remainder, client->_internal.registration_id);
   remainder = az_span_copy(remainder, provisioning_service_api_version);
 
   if (az_span_size(*user_agent) > 0)
   {
-    remainder = az_span_copy_u8(remainder, '&');
-    az_span_copy(remainder, *user_agent);
+    remainder = az_span_copy(remainder, user_agent_version_prefix);
+    remainder = az_span_copy(remainder, *user_agent);
   }
 
-  *out_mqtt_user_name = az_span_slice(mqtt_user_name, 0, required_length);
+  remainder = az_span_copy_u8(remainder, 0);
+
+  if (out_mqtt_user_name_length)
+  {
+    *out_mqtt_user_name_length = (size_t)required_length;
```

Comment:
nit: To be extra cautious, consider adding a precondition here that `_az_span_diff(remainder, mqtt_user_name_span) == required_length`

Reply:
I personally like having extra asserts to avoid bugs like the one @hihigupt pointed out above which would be detected by this kind of assert.
What I don't think is appropriate is
1. a "pre"condition which, I thought, were only to be used to verify parameters.
2. having code that cannot be tested in code coverage (if we do things right, this should never happen).

Reply:
I understand. I tend to re-use precondition also as something equivalent to .NET's `Debug.Assert` because they are closely related, which clearly isn't its purpose (like you said in 1). Usually code coverage not capturing `Debug.Assert` conditions is considered fine because it in itself is helping catch breaks in what a dev considered an invariant (for consistency in the internal logic) as the code evolves. 

@JeffreyRichter - what do you think about avoiding using the precondition for things outside of contract validation/parameter validation? Should that be avoided and discouraged, or does it not matter and we leave it to dev discretion?

Should we create a concept similar to `Debug.Assert` that extends precondition for more general purpose use?

Reply:
In the world of contracts there are preconditions, invariants, and post conditions, see https://en.wikipedia.org/wiki/Design_by_contract. A precondition (be definition) must be true prior to the execution of code. Typically, contracts are implemented via asserts. So, if we want some assert-like thing, that is fine, we can add it. But then, contracts would be implemented on top of it.

I think the PRECONDITION macros tell something to a customer and even our own developers so I like keeping this terminology/macro in our code at the top of function implementations. But, having an assert macro that does something similar is fine with me.

Reply:
In that case I won't add anything here. The check would validate if the code I wrote is correct. If it fails, there's nothing that the developer can do about it. I'll add UTs for this instead.

Reply:
https://github.com/Azure/azure-sdk-for-c/issues/638